File: Examples/iOS/Keyboard/Keyboard/Instruments/ToneGenerator.m
//
//  ToneGenerator.m
//  AudioKit Example
//
//  Created by Aurelius Prochazka on 6/9/12.
//  Copyright (c) 2012 Aurelius Prochazka. All rights reserved.
//

#import "ToneGenerator.h"

@implementation ToneGenerator

- (instancetype)init {
    self = [super init];
    if (self) {
        // NOTE BASED CONTROL ==================================================
        ToneGeneratorNote *note = [[ToneGeneratorNote alloc] init];
        [self addNoteProperty:note.frequency];
        [self addNoteProperty:note.amplitude];
        
        // INSTRUMENT CONTROL ==================================================
        _toneColor  = [[AKInstrumentProperty alloc] initWithValue:0.5
                                                          minimum:0.1
                                                          maximum:1.0];
        [self addProperty:_toneColor];
    
        
        // INSTRUMENT DEFINITION ===============================================
        
        AKLinearADSREnvelope  *adsr = [AKLinearADSREnvelope envelope];
        [self connect:adsr];
        
        AKFMOscillator *fmOscillator = [AKFMOscillator oscillator];
        fmOscillator.baseFrequency = note.frequency;
        fmOscillator.carrierMultiplier = [_toneColor scaledBy:akp(20)];
        fmOscillator.modulatingMultiplier = [_toneColor scaledBy:akp(12)];
        fmOscillator.modulationIndex = [_toneColor scaledBy:akp(12)];
        fmOscillator.amplitude = [adsr scaledBy:akp(0.15)];
        [self connect:fmOscillator];
        
        // EXTERNAL OUTPUTS ====================================================
        // After your instrument is set up, define outputs available to others
        _auxilliaryOutput = [AKAudio globalParameter];
        [self assignOutput:_auxilliaryOutput to:fmOscillator];
    }
    return self;
}

@end

// -----------------------------------------------------------------------------
#  pragma mark - Tone Generator Note
// -----------------------------------------------------------------------------


@implementation ToneGeneratorNote

- (instancetype)init {
    self = [super init];
    if (self) {
        _frequency = [[AKNoteProperty alloc] initWithMinimum:440 maximum:880];
        _amplitude = [[AKNoteProperty alloc] initWithMinimum:0 maximum:1];
        [self addProperty:_frequency];
        [self addProperty:_amplitude];
        _amplitude.value = 1.0;
    }
    return self;
}

- (instancetype)initWithFrequency:(float)frequency {
    self = [self init];
    if (self) {
        _frequency.value = frequency;
    }
    return self;
}

@end

File: Examples/iOS/SwiftKeyboard/SwiftKeyboard/Instruments/ToneGenerator.swift
//
//  ToneGenerator.swift
//  SwiftKeyboard
//
//  Created by Aurelius Prochazka on 11/28/14.
//  Copyright (c) 2014 AudioKit. All rights reserved.
//

class ToneGenerator: AKInstrument {

    // Instrument Properties
    var toneColor  = AKInstrumentProperty(value: 0.5, minimum: 0.1, maximum: 1.0)
    var auxilliaryOutput = AKAudio()
    
    override init() {
        super.init()

        // Instrument Properties
        addProperty(toneColor)

        // Note Properties
        let note = ToneGeneratorNote()
        addNoteProperty(note.frequency)
        addNoteProperty(note.amplitude)
        
        let adsr = AKLinearADSREnvelope()
        connect(adsr)
    
        let fmOscillator = AKFMOscillator()
        fmOscillator.baseFrequency = note.frequency
        fmOscillator.carrierMultiplier = toneColor.scaledBy(20.ak)
        fmOscillator.modulatingMultiplier = toneColor.scaledBy(12.ak)
        fmOscillator.modulationIndex = toneColor.scaledBy(15.ak)
        fmOscillator.amplitude = adsr.scaledBy(0.15.ak)
        connect(fmOscillator)
        
        auxilliaryOutput = AKAudio.globalParameter()
        assignOutput(auxilliaryOutput, to: fmOscillator)
    }
}

class ToneGeneratorNote: AKNote {
    
    // Note Properties
    var frequency = AKNoteProperty(minimum: 440, maximum: 880)
    var amplitude = AKNoteProperty(minimum: 0, maximum: 1)
    override init() {
        super.init()
        addProperty(frequency)
        addProperty(amplitude)
        amplitude.value = 1.0
    }
}